### PR TITLE
Fixed the crashing issue in Marathon by writing a JVM-only Zookeeper State...

### PR DESCRIPTION
**Pull request by swartzrock**
.. using the Apache Curator framework. The crash is avoided as JNI persistence is no longer used.   

https://github.com/mesosphere/marathon/issues/834

### CODE REVIEW

File: src/main/scala/mesosphere/util/JavaScalaFutureSupport.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  /**
+    * Converts a Scala Future to a Java Future
+    */
+  implicit class Scala2JavaFuture[A](f: Future[A]) {
+
+    def asJava(duration: Duration): JFuture[A] = {
+      val task = new FutureTask[A](new Callable[A] {
+        override def call(): A = Await.result(f, duration)
+      })
+      new Thread(task).run()
```
Comment (drexin) [inline src/main/scala/mesosphere/util/JavaScalaFutureSupport.scala]: This is very expensive. Please instead wrap the `scala.concurrent.Future` in a class that implements `java.util.concurrent.Future` and delegates the methods.

File: src/main/scala/mesosphere/util/ZookeeperDirectState.scala [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+  private val eventualCuratorClient: Future[CuratorFramework] = createZookeeperClient()
+
+  /**
+    * Block until we have a connection to Zookeeper. If the max time is reached
+    * an exception will be thrown.
+    * *
+    * @param duration the max amount of time to wait
+    */
+  def await(duration: Duration): Unit = {
```
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/util/ZookeeperDirectState.scala]: The method name isn't totally clear.  How about something along the lines of `awaitZookeeperConnection`?

**Comment by ConnorDoyle** [comment]
Awesome work @swartzrock!  It might make more sense to put this code in [mesos-utils](https://github.com/mesosphere/mesos-utils) so that other projects (e.g. Chronos) can also take advantage of the workaround.

File: src/main/scala/mesosphere/util/ZookeeperDirectState.scala [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+      sessionTimeout.toMillis.toInt,
+      connectionTimeout.toMillis.toInt,
+      RETRY_POLICY
+    )
+
+    client.start()
+
+    // Make sure we have a successful connection before returning the client
+    Future {
+      val connected = client.getZookeeperClient.blockUntilConnectedOrTimedOut()
```
Comment (drexin) [inline src/main/scala/mesosphere/util/ZookeeperDirectState.scala]: Wouldn't it be easier to create the `CuratorFramework` outside of this class and pass it in as parameter? Doing it in here makes the whole class more complex and blocks 2 threads.

File: src/main/scala/mesosphere/util/ZookeeperDirectState.scala [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      client
+    }
+  }
+
+  // Wrap our local Scala-Future-returning methods with Java-Future-returning methods
+  // to satisfy the base State interface.
+  override def names(): JFuture[util.Iterator[String]] = handleNames.asJava(sessionTimeout)
+  override def fetch(name: String): JFuture[Variable] = handleFetch(name).asJava(sessionTimeout)
+  override def store(v: Variable): JFuture[Variable] =
+    handleStore(v.asInstanceOf[LocalVariable]).asJava(sessionTimeout)
```
Comment (drexin) [inline src/main/scala/mesosphere/util/ZookeeperDirectState.scala]: No casting please.

File: src/main/scala/mesosphere/util/ZookeeperDirectState.scala [hunk 5]
```diff
@@ -0,0 +1,186 @@
+package mesosphere.util
+
+import java.lang.Boolean
+import java.util
+import java.util.UUID
+import java.util.concurrent.{ Future => JFuture }
+
+import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (drexin) [inline src/main/scala/mesosphere/util/ZookeeperDirectState.scala]: Please don't use this executor. We have one in `mesosphere.util.ThreadPoolContext`

File: src/main/proto/marathon.proto [hunk 6]
```diff
@@ -164,3 +164,17 @@ message TaskFailure {
...
+// Describes a state entry, a versioned (via a UUID) key/value pair.
+// Copied from libmesos (src/messages/state.pro) to ensure that Marathon's
+// persisted state is compatible with that of libmesos.
+message Entry {
+  required string name = 1;
+  required bytes uuid = 2;
+  required bytes value = 3;
+}
+
+
```
Comment (drexin) [inline src/main/proto/marathon.proto]: Why so many blank lines here?

File: src/main/proto/README.md [hunk 7]
```diff
@@ -5,6 +5,8 @@ Marathon is using Google [Protocol Buffers](https://developers.google.com/protoc
 ## Dependencies
 
 - Google [Protocol Buffers](https://developers.google.com/protocol-buffers) v2.5 should be installed on the system.
+-- Note: protoc 2.6 (and higher) will generate incompatible code. Make sure your version of protoc is 2.5.
```
Comment (drexin) [inline src/main/proto/README.md]: This change is unrelated to the PR, please create a separate PR for this as it would otherwise be lost if we ever have to revert this PR.

File: src/main/scala/mesosphere/util/ZookeeperDirectState.scala (hunk 4 above)
Comment (swartzrock) [inline src/main/scala/mesosphere/util/ZookeeperDirectState.scala]: Admittedly it doesn't look very good. However, this is replacing C++ code that performs the same typecast. Would it be better to remove the use of the JNI-based State and replace it with an implementation that won't require typecasting?

File: src/main/scala/mesosphere/util/ZookeeperDirectState.scala [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        .build()
+        .toByteArray
+
+      mkDir(Some(lv.name))
+      client.setData().forPath(fullPath(lv.name), entryData)
+      fetch(lv.name).get()
+    }
+    else {
+      log.warn(s"Attempt was made to persist ${lv.name} from obsolete version")
+      null
```
Comment (drexin) [inline src/main/scala/mesosphere/util/ZookeeperDirectState.scala]: This should be an exception instead.

File: src/main/scala/mesosphere/util/ZookeeperDirectState.scala (hunk 4 above)
Comment (drexin) [inline src/main/scala/mesosphere/util/ZookeeperDirectState.scala]: How about

```scala
val localVariable = v match {
  case lv: LocalVariable => lv
  case _ => // throw exception are something
}
```

**Comment by drexin** [comment]
@swartzrock Thanks for taking a shot at this. I added some comments inline. Overall I think there's way too much blocking here. Why don't we use the async Zookeeper client?

**Comment by guenter** [comment]
This is a pretty heavy handed workaround for a bug. We should fix the bug in the JNI bindings instead.

File: src/main/scala/mesosphere/util/JavaScalaFutureSupport.scala (hunk 1 above)
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/util/JavaScalaFutureSupport.scala]: We could also specialize the use site to grab the underlying Scala Future to avoid the call to `.get` in `BackToTheFuture`.

**Comment by swartzrock** [comment]
The core change to Marathon is that a different class is bound to State in MarathonModule. This does, of course, entail a change to its persistence logic, but mainly in the conversion from C++ to Scala.

**Comment by swartzrock** [comment]
Dario, Connor and Tobi - thank you for the awesome feedback! Tobi, re our discussion I will look into contributing a Java version of this (with all of the recommended fixes) to the org.apache.mesos Java library to reduce the Scala/Java Future conversions and make it more widely available to developers.
